pkg/github/repositories.go

```diff
@@ -1358,36 +1358,96 @@ func filterPaths(entries []*github.TreeEntry, path string, maxResults int) []str
 func resolveGitReference(ctx context.Context, githubClient *github.Client, owner, repo, ref, sha string) (*raw.ContentOpts, error) {
-	// 1. If SHA is provided, use it directly
+	// 1) If SHA explicitly provided, it's the highest priority.
 	if sha != "" {
 		return &raw.ContentOpts{Ref: "", SHA: sha}, nil
 	}
 
-	// 2. If neither provided, use the default branch as ref
+	// 2) If no SHA is provided, we try to resolve the ref into a fully-qualified format.
+	// 2a) If ref is empty, determine the default branch.
 	if ref == "" {
 		repoInfo, resp, err := githubClient.Repositories.Get(ctx, owner, repo)
 		if err != nil {
 			_, _ = ghErrors.NewGitHubAPIErrorToCtx(ctx, "failed to get repository info", resp, err)
 			return nil, fmt.Errorf("failed to get repository info: %w", err)
 		}
-		ref = fmt.Sprintf("refs/heads/%s", repoInfo.GetDefaultBranch())
+		ref = repoInfo.GetDefaultBranch()
 	}
 
-	// 3. Get the SHA from the ref
+	originalRef := ref // Keep original ref for clearer error messages down the line.
+
+	// Only enter the resolution logic if the ref is NOT already fully qualified.
+	if !strings.HasPrefix(ref, "refs/") {
+		if strings.HasPrefix(ref, "heads/") || strings.HasPrefix(ref, "tags/") {
+			// 2c) It's partially qualified. Make it fully qualified.
+			ref = "refs/" + ref
+		} else {
+			// 2d) It's a short name; try to resolve it as a branch or tag.
+			_, resp, err := githubClient.Git.GetRef(ctx, owner, repo, "refs/heads/"+ref)
+
+			if err == nil {
+				ref = "refs/heads/" + ref // It's a branch.
+			} else {
+				// The branch lookup failed. Check if it was a 404 Not Found error.
+				ghErr, isGhErr := err.(*github.ErrorResponse)
+				if isGhErr && ghErr.Response.StatusCode == http.StatusNotFound {
+					// The branch wasn't found, so try as a tag.
+					_, resp2, err2 := githubClient.Git.GetRef(ctx, owner, repo, "refs/tags/"+ref)
+					if err2 == nil {
+						ref = "refs/tags/" + ref // It's a tag.
+					} else {
+						// The tag lookup failed. Check if it was a 404 Not Found error.
+						ghErr2, isGhErr2 := err2.(*github.ErrorResponse)
+						if isGhErr2 && ghErr2.Response.StatusCode == http.StatusNotFound {
+							return nil, fmt.Errorf("could not resolve ref %q as a branch or a tag", originalRef)
+						}
+						// The tag lookup failed for a different reason.
+						_, _ = ghErrors.NewGitHubAPIErrorToCtx(ctx, "failed to get reference (tag)", resp2, err2)
+						return nil, fmt.Errorf("failed to get reference for tag '%s': %w", originalRef, err2)
+					}
+				} else {
+					// The branch lookup failed for a different reason.
+					_, _ = ghErrors.NewGitHubAPIErrorToCtx(ctx, "failed to get reference (branch)", resp, err)
+					return nil, fmt.Errorf("failed to get reference for branch '%s': %w", originalRef, err)
+				}
+			}
+		}
+	}
+
+	// Now that 'ref' is fully qualified, we get the definitive reference object.
 	reference, resp, err := githubClient.Git.GetRef(ctx, owner, repo, ref)
```

Comment:
It might be that you already retrieved a reference on line 1430 or 1439. In this case this call is redundant. We probably would want to avoid that.